Add support for language selection in preferences

goanpeca opened the pull request:
Fixes #1001 

---

## Description

Add support for language selection in preferences. 

This was a bit trickier than expected, namely the configuration of the language needs to be stored in a separate file, because we cannot use CONF at the stage the `_` function is called.

I added a `langconfig` in the .`spyder` folder, which stores the language value ('en', 'fr' etc)

With this change I had to implement some new logic in the config dialog, so now comboboxes (so far) have an extra parameter call `restart`  (by default False).

## Preferences dialog

![image](https://cloud.githubusercontent.com/assets/3627835/7445768/bbff349a-f1c0-11e4-940e-9a745e7df4e7.png)


## Pop up message box
When the user changes one or more options flagged with `restart=True` and then clicks on Apply or Ok, the a message box will popup informing the user which options need to restart and asking if he/she wants to restart right away.

![image](https://cloud.githubusercontent.com/assets/3627835/7442847/c00155bc-f122-11e4-98a9-b6952ceae065.png)

If the user accepts, any restart specific (defined) settings will be applied and spyder will restart automatically.

### Note:
The message box is now available for any other combobox that is created with the `restart=True` parameter. Adding this functionality to other widgets will be left for another PR.

## Todo
- [x] Display language names In original Language
- [x] Agree on additional file method and name
- [x] Define when to pop the box to the user for restart

## For another PR
- Add the restart functionality to all the other widgets besides comboboxes

ccordoba12 commented on spyderlib/baseconfig.py:
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
 #==============================================================================
 # Translations
 #==============================================================================
+LANG_FILE = osp.join(get_conf_path(), '.langconfig')
```
I'm ok with adding this new file, but don't call it `.langconfig`, just `langconfig`, so that it's not hidden on Linux :-)

At the end, it's going to be inside of `~/.spyder`, so there is no need to be hidden.

goanpeca replied:
:+1:

ccordoba12 commented on spyderlib/baseconfig.py:
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
+LANGUAGE_CODES = {'en': u'English',
+                  'fr': u'Français',
+                  'es': u'Español',
+                  'pt_BR': u'Português (brasileiro)'
```
Is Brazilian Portuguese too different from Portuguese? If not, I'd vote to just leave Portuguese here

goanpeca replied:
We should ask the guy that did the translation. Can you add him to the thread?

goanpeca replied:
But i do prefer to put just portuguese...

goanpeca commented on spyderlib/plugins/configdialog.py:
```diff
@@ -192,6 +199,7 @@ def current_page_changed(self, index):
         self.apply_btn.setEnabled(widget.is_modified)
         
     def add_page(self, widget):
+        setattr(widget, 'parent_config_dialog', self)
```
Oops i do no longer need this. Will remove it.

ccordoba12 commented on spyderlib/baseconfig.py:
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
+                  }
+
+
+def get_available_translations():
```
I don't think we need a function for this ;-)

There are so few translations (and none on the works), that we can just create the list of `langs` by hand here.

goanpeca replied:
Probably we do not need it now, but it is a harmless function, and it provides a check when adding a new language to make sure things are done properly. I think we should leave it

ccordoba12 replied:
Ok, you're right :-)

If you want more translations, you should look at https://www.transifex.com/

I don't know exactly how it works, but I think we need to upload our pot files and volunteers help to translate them to their native languages :-)

goanpeca replied:
I will take a look a transifex, I had seen it before, but never got the time to go in depth

ccordoba12 commented on spyderlib/baseconfig.py:
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
+    return langs
+
+
+def get_interface_language():
```
It also seems like too much to have this function

goanpeca replied:
Yes it might be too much, but why doing stuff by hand, if a function can handle it?

I would really like to have many more translations

ccordoba12 replied:
Ok, now that it's written, we should leave it  :-)

ccordoba12 commented on spyderlib/utils/iofuncs.py:
```diff
@@ -362,7 +362,8 @@ def load_dictionary(filename):
                       '.projects', '.spyderproject', '.ropeproject',
-                      'monitor.log', 'monitor_debug.log', 'rope.log')
+                      'monitor.log', 'monitor_debug.log', 'rope.log',
+                      '.langconfig')
```
`.langconfig` -> `langconfig`

goanpeca replied:
Sure

goanpeca commented:
@ccordoba12 I updated with the fixes, when you have some time can you take a look at some questions I left in the description?

goanpeca commented:
Can @spyder-ide/core-developers give this a try and give some comments?

```bash
git clone https://github.com/goanpeca/spyder.git spyder-lang
cd spyder-lang
git checkout -b language
git pull origin language
python bootstrap.py
```

goanpeca commented:
This PR is waiting for some input from the @spyder-ide/core-developers ...

There are 2 ways of doing this.
1.) Ask the user if he wants to restart for the setting to take place, immediately after changing the setting
2.) Ask the user if he wants to restart for the setting to take place, only after pressing Ok or Apply

I implemented the first one.. but the second one could be also implemented if you consider is better.

@ccordoba12 any ideas, thoughts?

Nodd commented:
I vote 2 because:
* there could be other settings which ask to reload (changing the icon theme, enabling plugins…)
* It's annoying to have a popup without being able to check if the change was correct.

If 2 is chosen, it should be written that it requires a restart.

goanpeca commented:
In that case of 2, I would make the popup appear at the end when clicking `ok`, or `apply`

Nodd commented:
Yep. Bonus points if the dialog contains the reason why the restart is needed (as a list so that other settings can give their information too)

goanpeca commented:
Yep, you read my mind ;-)

goanpeca commented:
@Nodd ok, this is now updated to ask only after pressing Ok or Apply and displays a list of all the options that require a restart (and were changed).

Can you give it a try @ccordoba12? 

I would expect this to work in mac without changes but, @blink1073 can you give it a try?

ccordoba12 commented:
The restart popup message is very nice!! It looks very professional! Thanks a lot for working on it @goanpeca!

ccordoba12 commented on spyderlib/baseconfig.py:
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
+        language = DEFAULT_LANGUAGE
+    else:
+        spyder_languages = get_available_translations()
+        for spyder_language in spyder_languages:
```
Let's simplify the variables here to make it a bit easier to read. What about this?

```python
for lang in spyder_languages
```

goanpeca replied:
Ok sure, I will adapt ;-)

goanpeca commented:
Glad you like it, if this works on Mac, then it should be ready to merge

ccordoba12 commented on spyderlib/baseconfig.py:
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
+            if locale_language == spyder_language:
+                language = locale_language
+                break
+            elif locale_language.startswith(spyder_language) or\
```
Missing space: `or\` -> `or \`

goanpeca replied:
will fix, but probably will be one line with your suggestion

ccordoba12 commented on spyderlib/baseconfig.py:
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
+                language = locale_language
+                break
+            elif locale_language.startswith(spyder_language) or\
+                    spyder_language.startswith(locale_language):
```
2 indentation spaces here, not 8, please (I know PEP8, but I don't like it)

ccordoba12 commented on spyderlib/baseconfig.py:
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
+
+    # fixup environment var LANG, LANGUAGE
+    language = load_lang_conf()
+    os.environ["LANG"] = language      # Works in windows
```
`works in windows` -> `Works on Windows`

The same for Ubuntu below, but changing it to Linux :-)

goanpeca replied:
Is this going to work on Mac for sure?

ccordoba12 replied:
Translations on Mac never have worked because Mac doesn't define proper `locale` values.

This work will let users select the translation they want to use :-)

ccordoba12 commented on spyderlib/plugins/configdialog.py:
```diff
@@ -90,6 +91,18 @@ def apply_changes(self):
+            for restart_option in self.restart_options:
+                if restart_option in self.changed_options:
+                    self.prompt_restart_required()
+                    break
```
Why this `break` here?

goanpeca replied:
I guess that I am under the impression that if I do not break it here, you will get a popup for every option that the user changed (and requires restart)

ccordoba12 replied:
I see, it's fine then :-)

ccordoba12 commented on spyderlib/plugins/configdialog.py:
```diff
@@ -661,13 +680,42 @@ def get_icon(self):
+    def prompt_restart_required(self):
+        """Prompt the user with a request to restart."""
+        restart_opt = self.restart_options
+        changed_opt = self.changed_options
```
`restart_opt` -> `restart_opts` (because it's plural)
`changed_opt` -> `changed_opts`

goanpeca replied:
fixed

ccordoba12 commented on spyderlib/plugins/configdialog.py:
```diff
@@ -704,6 +752,13 @@ def setup_page(self):
 
+        # Language chooser
+        langs = get_available_translations()
+        choices = list(zip([LANGUAGE_CODES[key] for key in langs], langs))
```
What this `choices` variable hold? I mean, is it something like

```python
[('en': 'English', 'es': 'Spanish')]
```

?

goanpeca replied:
I copied this from code above (the choices part and adapted it...) 

It holds what is stored in `LANGUAGE_CODES` in the `baseconfig.py` 

So the combo box display Español, English... etc... and the value is ...es, en... etc..

goanpeca replied:
So actually it is more like 

`choices = [(English, 'en'), ('Español', 'es'), ...]`

ccordoba12 replied:
This seems a bit too complex to my taste (I mean, the difference between `LANGUAGE_CODES` and `get_available_translations`).

What will happen when someone adds `pt_PT` or `en_UK`?

goanpeca replied:
When someone adds this where?

goanpeca replied:
I could just use the `LANGUAGE_CODES` variable and remove the call to `langs = get_available_translations()`

ccordoba12 commented on spyderlib/baseconfig.py:
```diff
@@ -204,18 +205,102 @@ def get_image_path(name, default="not_found.png"):
 #==============================================================================
 # Translations
 #==============================================================================
+LANG_FILE = osp.join(get_conf_path(), 'langconfig')
```
Please rewrite this line as

```python
LANG_FILE = get_conf_path('langconfig')
```

That's the way `get_conf_path` works :-)

goanpeca replied:
Done

ccordoba12 commented:
Everything seems fine to me now, but:

1. I still find the implementation too complex to follow.
2. We still need to test this PR on Mac :-)

goanpeca commented:
Regarding 1.) hmmm

I tried to add all the needed documentation to understand. I will explain and you tell me where you find it confusing.

When someone wants to add a new translation they will create the folder in `locale` and put there the new translation, or variation of a translation.

Once they do this and run the code they will get an error because we need to have an updated language name (in native language) added to `LANGUAGE_CODES`. If they added a new language, no problem they just add i.e. `'de': 'Deutsch'`... if it is a variation they just add i.e. `'pt_PT': 'Portuguese (Portugal)'` and make sure the other variation (if existent) should be then `'pt_BR': 'Portuguese (Brazil)'`

But the truth is, that the second case will most likely never happen...

goanpeca commented on spyderlib/plugins/configdialog.py:
```diff
@@ -704,6 +752,12 @@ def setup_page(self):
+        language_combo = self.create_combobox(_('Language'), choices,
+                                              'interface_language',
+                                              restart=True)
+
```
@ccordoba12 this should be clearer now

ccordoba12 replied:
Yep, much better, thanks! Now I don't have objecttions.

goanpeca replied:
:+1:

ccordoba12 commented:
I'll merge this tomorrow, after testing on Mac :-)

goanpeca commented:
I will have to buy a Mac... ;-)

blink1073 commented:
I can test today on my Mac if you prefer @ccordoba12.

ccordoba12 commented:
That'd be great too @blink1073! I thought you weren't around ;-)

blink1073 commented:
Was at the beach with the little ones.  ;-)

blink1073 commented:
Una problema:

```python
Cargando el inspector de objetos...
Traceback (most recent call last):
  File "/Users/silvester/workspace/spyder/spyderlib/spyder.py", line 3077, in main
    mainwindow = run_spyder(app, options, args)
  File "/Users/silvester/workspace/spyder/spyderlib/spyder.py", line 2963, in run_spyder
    main.setup()
  File "/Users/silvester/workspace/spyder/spyderlib/spyder.py", line 851, in setup
    self.set_splash(_("Loading outline explorer..."))
  File "/Users/silvester/workspace/spyder/spyderlib/spyder.py", line 2095, in set_splash
    self.debug_print(message)
  File "/Users/silvester/workspace/spyder/spyderlib/spyder.py", line 532, in debug_print
    debug_print(message)
  File "/Users/silvester/workspace/spyder/spyderlib/baseconfig.py", line 56, in debug_print
    print(*message, file=ss)
UnicodeEncodeError: 'ascii' codec can't encode character '\xf3' in position 27: ordinal not in range(128)
```

blink1073 commented:
It looks like `sys.stdout` is encoded as `ascii`, thanks Apple...

goanpeca commented:
Any suggestions?

blink1073 commented:
Looking into it now.

blink1073 commented:
Hmm, this works fine from an IPython prompt:  `print('€', file=sys.stdout)`

blink1073 commented:
It might have something to do with the restart itself affecting `sys.stdout`.

goanpeca commented:
If you say No when the popup box asks to restart... and then you close spyder...
and rerun from bootstrap...?

You get the same error?

blink1073 commented:
If I restart after the failure itself it comes up fine.

goanpeca commented:
If you run without DEBUG ?

blink1073 commented:
Actually if I say NO then it restarts anyway...

goanpeca commented:
hmm thats a bug then.. it should do that!

blink1073 commented:
Oops, forget that last comment.

goanpeca commented:
Can you run without Debug? and check again?

blink1073 commented:
Running without debug is fine, and I can print unicode characters from the Spyder Python console.

blink1073 commented:
It appears to just be the restart while in debug mode.

blink1073 commented:
I tried using `sys.stdout` within `debug_print`, instead of the global variable, but it still did not work.

goanpeca commented:
Can you try this...

Change baseconfig.py  line 52 with this

```python
def debug_print(*message):
    """Output debug messages to stdout"""
    if DEBUG:
        ss = STDOUT
        message = tuple([m.encode('utf-8') for m in message])
        print(*message, file=ss)
```

And try with debug again?

blink1073 commented:
That works but all of the strings print with `b''`

blink1073 commented:
E.g.: `b'Configuration de la fen\xc3\xaatre principale...'`

goanpeca commented:
and if you do this

```python
def debug_print(*message):
    """Output debug messages to stdout"""
    if DEBUG:
        ss = STDOUT
        message = tuple([to_text_string(m) for m in message])
        print(*message, file=ss)
```

blink1073 commented:
Nope, same UnicodeEncodeError.

goanpeca commented:
You are on Py3?

blink1073 commented:
Yep, I'll try Py2.

blink1073 commented:
Same problem there.

blink1073 commented:
If I turn DEBUG off, do the restart, and then try and print from with the Spyder Internal Console, I get the same error too.  The other consoles are note affected.

goanpeca commented:
yes... the internal console seems to be the problem...

But that is something unrelated to the changes I did, seems to be a problem that has been sleeping for some time. I am trying some options

blink1073 commented:
Yes, it appears to be related to the restart itself.

blink1073 commented:
Yup, confirmed.  If I start in English mode and do just a restart, we get this behavior.

goanpeca commented:
I will try linux.. one sec

goanpeca commented:
Ok, I can reproduce it, ...fixing!! give me 10 mins

goanpeca commented:
@blink1073 

Can you add this line to `spyderlib\spyder.py` after line 2737

`env['PYTHONIOENCODING'] = 'utf-8'`

blink1073 commented:
Perfecto!

goanpeca commented:
both py2 and py3?

blink1073 commented:
Oui.

blink1073 commented:
`debug_print` is super nice, I use it all the time.

Nodd commented:
This env variable will change the behavior in scripts run from spyder ? This is bad, we had many bug reports because of this. We should change the encoding of the internal console only. I think I have a solution somewhere, I'll try to find it.

goanpeca commented:
@Nodd I do not think it will be the case... but you can try with an example?

goanpeca commented:
I can erase the environment variable in the `restart_app.py` just to be sure... so it is only used for the python process that spawns the new Spyder instance.

goanpeca commented:
Maybe I can pop that variable in `spyderlib\spyder.py` so that it is not there after startup. It is needed when spawning the new Spyder process on restart, but should not be needed afterwards.

blink1073 commented:
But what if the user had explicitly set that variable to something else?  Perhaps only set it if it is not in `os.environ` and not set to something else, then leave it there.

blink1073 commented:
`utf-8` all the things!

goanpeca commented:
Well, if the user set it to soemthing else, the spyder will break...

blink1073 commented:
Then I say set it and leave it.

goanpeca commented:
Lets see what @Nodd brings :-)

SylvainCorlay commented:
I am good with this. Will use that dialog on restarting spyder for the icons as soon as this is merged.

goanpeca commented:
@Nodd did you have some time to find what you were looking for? [or](https://www.youtube.com/watch?v=EabB55IYyOg)

@ccordoba12  any comments on the fix needed to get this PR working properly, specifically [this last commit](https://github.com/goanpeca/spyder/commit/5bae10be0773ab2a3b3a2a943606ab49e4d69d96)

Nodd commented:
I used this in some scritps so that `argparse` help doesn't raise a `UnicodeSomethingError` when displaying a help message with utf8 symbols under python2:
```
import codecs
sys.stdout = codecs.getwriter('utf-8')(sys.stdout)
```
Since the internal console is basically `sys.stdout`, it might be used as-is.

goanpeca commented:
and in python 3 ?

goanpeca commented:
where should I put this code?

Nodd commented:
In python3 I didn't have this problem (the unicode support is way better).
It should be put at the same place where sys.stdout is redirected to the internal console, I guess. I have no idea where it is :disappointed:

ccordoba12 commented on spyderlib/spyder.py:
```diff
@@ -2743,6 +2743,9 @@ def restart(self):
         else:
+            # This environment variable is needed so that the new python
+            # process spawned with Popen sets this as the default encoding
+            env['PYTHONIOENCODING'] = 'utf-8'
```
I think setting this system-wide is not a good idea. For example, It could mess with Pandas trying to import/export files :-)

ccordoba12 commented:
If I'm not wrong, redirection of stdio for the internal console happens in `redirect_internalshell_stdio`, currently present in

https://github.com/spyder-ide/spyder/blob/master/spyderlib/spyder.py#L2435

goanpeca commented:
I tried putting @Nodd  where @ccordoba12 without success

goanpeca commented:
@Nodd I think I have it fixed, can you check the latest commit?

@blink1073, or @ccordoba12  can you please test on Mac? remember to run using 

`python bootstrap.py --debug`

blink1073 commented:
It works for me, other than the fact that the restart dialog was entirely in English.

blink1073 commented:
![screen shot 2015-05-07 at 8 26 03 pm](https://cloud.githubusercontent.com/assets/2096628/7529020/58dea1c2-f4f7-11e4-9d5c-b05f817f30e8.png)

goanpeca commented:
Cause we need To rerun the localización scripts :+1:

goanpeca commented:
@ccordoba12 can we merge this one so that @SylvainCorlay can add this to his icon PR :smile: ?

SylvainCorlay commented:
:+1:

SylvainCorlay commented:
Yeah seems I will need to rebase again anyway!

ccordoba12 commented on spyderlib/config.py:
```diff
@@ -169,6 +169,7 @@ def is_ubuntu():
               'animated_docks': True,
               'prompt_on_exit': False,
               'panes_locked': True,
+              'interface_language': load_lang_conf(),
```
I think we don't need to add this here because this is dependent on the user system.

The only options that we add in `config.py` are those that can be changed and/or updated in the future (by bumping `CONF_VERSION`). That's why (for example) we don't add here the window size or state.

ccordoba12 replied:
Another example of things we don't add are the run configuration options per file :-)

ccordoba12 replied:
Please remove this option from here to finally merge this PR :-)

If I understand the code correctly:

1. It'll be saved to `spyder.ini` by the `save_lang` method of `MainConfigPage`
2. It can't be retrieved by `CONF` directly.

So there's no need to have it here.

goanpeca replied:
@ccordoba12 

If I don't add that method to the config, Spyder will not be able to load the preferences after a reset, cause there will be no `interface_language` setting in the .ini file.

Suggestions?

goanpeca replied:
I think we should leave it as it is, otherwise we would have to made unneded changes either to how the preferences are loaded or do an extra check in spyder.y to see if this is the first time it runs.... here it is just one line of code...

ccordoba12 commented on spyderlib/plugins/configdialog.py:
```diff
@@ -90,8 +91,20 @@ def apply_changes(self):
+            # is needed before loading CONF, this is an extra method needed to
+            # ensure that when changes are applied, they are copied to a
+            # specific file storing the language value.
+            if getattr(self, 'save_lang', False):
```
Why is this line necessary?

goanpeca replied:
Cause this class is inherited by all config panes. But only the main will for sure have the save_lang method.

Nodd replied:
how about changing the name to `__save_lang` ? this way it will be "hidden" to the child class (external plugins included), but the name will be changed at runtime (the class name is added).

goanpeca replied:
sure

SylvainCorlay commented:
Is there any blocking point for this one?

goanpeca commented on spyderlib/config.py:
```diff
@@ -724,6 +724,9 @@ def is_ubuntu():
                   subfolder=SUBFOLDER, backup=True, raw_mode=True)
 
+# Ensures that the config is present on spyder first run
+CONF.set('main', 'interface_language', load_lang_conf())
```
@ccordoba12 based on your previous comment I need to add this line here to make sure that on first run of spyder the `interface_language` entry is created. On subsequent starts of spyder that line is harmless so this way is not part of the CONF but ensures the correct behavior. 

You think is ok like this?

ccordoba12 replied:
Yes, I think it's fine :-) But I think it's better to move it `configdialog.py`. Well, if it's possible because there is where this setting is really needed.

goanpeca replied:
Hmmm ok I will take a look if I find a suitable place there

goanpeca commented:
@ccordoba12 this one should be ready to merge as well

SylvainCorlay commented:
:+1:

ccordoba12 commented on spyderlib/plugins/configdialog.py:
```diff
@@ -90,8 +91,21 @@ def apply_changes(self):
+            # ensure that when changes are applied, they are copied to a
+            # specific file storing the language value. This only applies to
+            # the main section config.
+            if self.CONF_SECTION == u'main':
```
I don't understand this change here. Maybe it's better to just move the `CONF.set` line from `config.py` here to avoid this.

ccordoba12 replied:
I mean, what I don't understand is why you're looking for `main`.

goanpeca replied:
`main` is the section in charge of setting the interface language, so the `_save_lang` only belongs to that section, before I was using  `if getattr(self, "_save_lang", False):` but I change it so that it was more explicit.

Since this method runs for every child (any other section or plugin) then I need this to avoid an error of calling _save_lang when it does not exist.

goanpeca replied:
The other way would be to make the save_lang method part of one of the ancestors so that EVERY section would alwyas make that save... but since it is only necessary for main...

ccordoba12 replied:
Too much work :-) Then leave this part as it is now.

Please see if you can move the `CONF.set` call from `config.py` to here so that we can (finally!) merge this PR ;-)

goanpeca replied:
Will do! ;-)

goanpeca commented on spyderlib/plugins/configdialog.py:
```diff
@@ -151,6 +165,9 @@ def __init__(self, parent=None):
+
+        # Ensures that the config is present on spyder first run
+        CONF.set('main', 'interface_language', load_lang_conf())
         
```
Moved here, right at the end of the init :-), should work fine now

goanpeca replied:
@ccordoba12

goanpeca commented:
Merging :-) @ccordoba12?

ccordoba12 commented:
Yes!! I'm having a problem with Qt5 but it seems to be with my system rather than with this PR (because I'm also having problems in master :-)

ccordoba12 commented:
Thanks for your patience by the way ;-)

goanpeca commented:
No prob ;-)

Ok, @SylvainCorlay, now is your turn :-)  

Just add `restart=True` to the combobox, i.e.

```python
        language_combo = self.create_combobox(_('Language'), choices,
                                              'interface_language',
                                              restart=True)
```
